docs/Implementing-easyblocks.rst

```diff
@@ -0,0 +1,486 @@
+
+This documentation covers aspects of implementing *easyblocks* and how to use them.
+For a concise definition of easyblocks, see :ref:`easyblocks`.
+
+.. contents::
+    :depth: 3
+    :backlinks: none
+
+
+.. _implementing_easyblocks_basics:
+
+The basics
+----------
+
+An *easyblock* is a Python module that implements a software build and installation procedure.
+
+It can be considered a Python script that plugs into the :ref:`framework`.
+
+EasyBuild will leverage easyblocks as needed, depending on which software packages it needs to install.
+Which easyblock is required is determined by the ``easyblock`` easyconfig parameter, or the software name
+(see :ref:`implementing_easyblocks_generic_vs_software_specific`).
+
+
+.. _implementing_easyblocks_generic_vs_software_specific:
+
+Generic vs software-specific easyblocks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Easyblocks can either be *generic* of *software-specific*.
+
+Generic easyblocks implement a 'standard' software build and installation procedure that is used by multiple different
+software packages.
+A commonly used example is the
+`ConfigureMake <https://github.com/hpcugent/easybuild-easyblocks/blob/master/easybuild/easyblocks/generic/configuremake.py>`_ 
+generic easyblock, which implements the standard ``configure`` - ``make`` - ``make install`` installation procedure used
+by most GNU software packages. See also :ref:`generic_easyblocks`.
+
+Software-specific easyblocks implement the build and installation procedure for a particular software package.
+Typically this involves highly customised substeps, for example specyfing dedicated configuration options, creating
+or adjusting specific files, executing non-standard shell commands, etc. Usually a custom implementation of the
+sanity check is also included.
+
+Since EasyBuild v2.0, using a generic easyblock requires specifying the ``easyblock`` parameter in the easyconfig file;
+if it is not specified, EasyBuild will try and find the software-specific easyblock derived from the software name.
+
+The distinction between generic and software-specific easyblocks can be made based on the naming scheme that is used
+for an easyblock, see also :ref:`implementing_easyblocks_naming_scheme`.
+
+
+.. _implementing_easyblocks_vs_easyconfigs:
+
+Easyblocks vs easyconfigs
+-------------------------
+
+Before you start implementing an easyblock, you should determine whether or not you really need an easyblock.
+
+:ref:`easyconfig_files` provide quite a bit of flexibility that alleviate the need for implementing many (software-specific)
+easyblocks. Examples include easyconfig parameters like ``(pre){config,build,installopts}`` (available for any easyblock),
+``install_cmd`` (only for ``Binary`` generic easyblock or derivates), etc. See :ref:`writing_easyconfig_files` for an
+overview of what is supported in easyconfig files, and :ref:`generic_easyblocks` for custom easyconfig parameters for
+the existing generic easyblocks.
```

Comment:
Isn't there a command arg to ask for all the parameters available for a particular easyblock?